### PR TITLE
cocoadisplay: Use hotspot read from .cur files

### LINKED ISSUE #845: OSX mouse "hotspot" in .cur file not being updated
When `.setCursorFilename("cursorFile.cur")` is assigned, the value of the "hotspot" on Windows is carried properly, but in OSX is not updated, leaving the "hotspot" in the top right position.

Have not tested it in Linux.

```python
wp = WindowProperties()
wp.setCursorFilename("crossHair.cur")
base.win.requestProperties(wp)
```

added .cur file with modified hotspot in zip file.

[crossHair.zip](https://github.com/panda3d/panda3d/files/4056632/crossHair.zip)

### PR DESCRIPTION
**Pull request by treamology**
Tested this with setting the cursor through WindowProperties and through a .prc file. Any non .cur file should default to the origin.

Fixes #845

### CODE REVIEW

File: panda/src/cocoadisplay/cocoaGraphicsWindow.mm [hunk 1]
```diff
@@ -568,20 +568,45 @@
   }
 
   if (_properties.has_cursor_filename()) {
-    NSImage *image = load_image(_properties.get_cursor_filename());
+    NSData *image_data = load_image_data(_properties.get_cursor_filename());
+
+    // Read the metadata from the image, which should contain hotspotX and
+    // hotspotY properties.
+    CGImageSourceRef cg_image = CGImageSourceCreateWithData((CFDataRef)image_data, nullptr);
+    NSDictionary *image_props = (NSDictionary *)CGImageSourceCopyPropertiesAtIndex(cg_image, 0, nil);
```
Comment (rdb) [inline panda/src/cocoadisplay/cocoaGraphicsWindow.mm]: This is a copy operation, yet the properties object is not being released.

To be on the safe side, perhaps we can also check that `image_props` isn't null before accessing it.  I've been bitten before by this kind of thing.

File: panda/src/cocoadisplay/cocoaGraphicsWindow.mm [hunk 2]
```diff
@@ -568,20 +568,45 @@
...
+      }
       _cursor = cursor;
     } else {
       _properties.clear_cursor_filename();
     }
     // This will ensure that NSView's resetCursorRects gets called, which sets
     // the appropriate cursor rects.
     [[_view window] invalidateCursorRectsForView:_view];
+
+    [image release];
```
Comment (rdb) [inline panda/src/cocoadisplay/cocoaGraphicsWindow.mm]: Should this be inside a nil check?

File: panda/src/cocoadisplay/cocoaGraphicsWindow.mm [hunk 3]
```diff
@@ -1031,10 +1056,31 @@
       properties.set_cursor_filename(cursor_filename);
       properties.clear_cursor_filename();
     } else {
-      NSImage *image = load_image(cursor_filename);
+      NSData *image_data = load_image_data(cursor_filename);
+
+      // Read the metadata from the image, which should contain hotspotX and
+      // hotspotY properties.
+      CGImageSourceRef cg_image = CGImageSourceCreateWithData((CFDataRef)image_data, nullptr);
+      NSDictionary *image_props = (NSDictionary *)CGImageSourceCopyPropertiesAtIndex(cg_image, 0, nil);
```
Comment (rdb) [inline panda/src/cocoadisplay/cocoaGraphicsWindow.mm]: This code is duplicated, perhaps we could create a `load_cursor` method that returned an NSCursor pointer?  Or, failing that, a load_image_data that took the CGFloat hotspot_x/y as reference arguments?

File: panda/src/cocoadisplay/cocoaGraphicsWindow.mm [hunk 4]
```diff
@@ -1287,18 +1335,24 @@
...
     return nil;
   }
 
-  image = [[NSImage alloc] initWithData:data];
-  [data release];
+  _images[resolved] = data;
+  return data;
+}
+
+NSImage *CocoaGraphicsWindow::
```
Comment (rdb) [inline panda/src/cocoadisplay/cocoaGraphicsWindow.mm]: This method is missing a comment block.

File: panda/src/cocoadisplay/cocoaGraphicsWindow.mm [hunk 5]
```diff
@@ -568,20 +568,45 @@
   }
 
   if (_properties.has_cursor_filename()) {
-    NSImage *image = load_image(_properties.get_cursor_filename());
+    NSData *image_data = load_image_data(_properties.get_cursor_filename());
+
+    // Read the metadata from the image, which should contain hotspotX and
+    // hotspotY properties.
+    CGImageSourceRef cg_image = CGImageSourceCreateWithData((CFDataRef)image_data, nullptr);
```
Comment (rdb) [inline panda/src/cocoadisplay/cocoaGraphicsWindow.mm]: Oh, and `load_image_data` may return null, but if it does, this call will probably crash.

**Comment by el-dee** [comment]
It seems this does not compile with macOS SDK 10.6 : 
<pre>
In file included from panda/src/cocoadisplay/p3cocoadisplay_composite1.mm:5:
panda/src/cocoadisplay/cocoaGraphicsWindow.mm:1424:7: error: expected method to read dictionary element not found on object of type 'NSDictionary *'
  if (image_props[@"hotspotX"] != nil) {
      ^
panda/src/cocoadisplay/cocoaGraphicsWindow.mm:1425:30: error: expected method to read dictionary element not found on object of type 'NSDictionary *'
    hotspot_x = [(NSNumber *)image_props[@"hotspotX"] floatValue];
                             ^
panda/src/cocoadisplay/cocoaGraphicsWindow.mm:1427:7: error: expected method to read dictionary element not found on object of type 'NSDictionary *'
  if (image_props[@"hotspotY"] != nil) {
      ^
panda/src/cocoadisplay/cocoaGraphicsWindow.mm:1428:30: error: expected method to read dictionary element not found on object of type 'NSDictionary *'
    hotspot_y = [(NSNumber *)image_props[@"hotspotY"] floatValue];
</pre>

**Comment by rdb** [comment]
Yeah, I just noticed the buildbot failure.  Would either of [valueForKey](https://developer.apple.com/documentation/foundation/nsdictionary/1410210-valueforkey?language=objc) or [objectForKey](https://developer.apple.com/documentation/foundation/nsdictionary/1414347-objectforkey?language=objc) work instead?  We can leave the syntactic-sugar-version on the master branch.

**Comment by treamology** [comment]
`objectForKey` would be what we would want to use, `valueForKey` is meant for key-value coding on any NSObject.

**Comment by rdb** [comment]
Should be fixed by e011f88f52f099658b65d1368d84f6217145ba0f.  That commit also fixes an issue with `cursor` being uninitialized if `image` is `nil` (which I missed during review).

**Pull request closed without merging**